lib/nnc/cmd/blas/mps/ccv_nnc_mul_mps.m

```diff
@@ -87,6 +87,171 @@ static int _ccv_nnc_mul_forw(const ccv_nnc_cmd_t cmd, const ccv_nnc_hint_t hint,
+		MPSGraphExecutable* executable = ccv_nnc_mps_graph_executable_cache(key, indices, ^void (MPSGraph* graph, NSMutableArray<MPSGraphTensor*>* inputTensors, NSMutableArray<MPSGraphShapedType*>* inputShapedTypes, NSMutableArray<MPSGraphTensor*>* resultTensors) {			
+			MPSGraphTensor* mps_input_b1;
+			MPSGraphTensor* mps_b1 = ccv_nnc_mps_graph_tensor_input(graph, b1, b1->info.dim, b1->stride, &mps_input_b1);
+			[inputTensors addObject:mps_input_b1];
+			MPSGraphShapedType* mps_b1_shape = ccv_nnc_mps_graph_tensor_input_shape(b1, b1->info.dim, b1->stride);
+			[inputShapedTypes addObject:mps_b1_shape];
+
+			MPSGraphTensor* mps_input_b2;
+			MPSGraphTensor* mps_b2 = ccv_nnc_mps_graph_tensor_input(graph, b2, b2->info.dim, b2->stride, &mps_input_b2);
+			[inputTensors addObject:mps_input_b2];
+			MPSGraphShapedType* mps_b2_shape = ccv_nnc_mps_graph_tensor_input_shape(b2, b2->info.dim, b2->stride);
+			[inputShapedTypes addObject:mps_b2_shape];
+			
+			if (g) {
+				MPSGraphTensor* mps_input_g;
+				MPSGraphTensor* mps_g = ccv_nnc_mps_graph_tensor_input(graph, g, g->info.dim, g->stride, &mps_input_g);
+				[inputTensors addObject:mps_input_g];
+				MPSGraphShapedType* mps_g_shape = ccv_nnc_mps_graph_tensor_input_shape(g, g->info.dim, g->stride);
+				[inputShapedTypes addObject:mps_g_shape];
+
+			MPSGraphTensor* mps_a = mps_g;
+			if (p != 1)
+			{
+				MPSGraphTensor* mps_p = [graph constantWithScalar:p dataType:ccv_nnc_mps_datatype(g->info.datatype)];
+				mps_a = [graph multiplicationWithPrimaryTensor:mps_a secondaryTensor:mps_p name:nil];
+
+			}
+			mps_a = [graph multiplicationWithPrimaryTensor:mps_a secondaryTensor:mps_b1 name:nil];
+
+			NSMutableArray<NSNumber*>* da_axes = [NSMutableArray new];
+			const int a_nd = ccv_nnc_tensor_nd(a->info.dim);
+			for (int i = 0; i < a_nd; i++) {
+				if (a->info.dim[i] != g->info.dim[i])
+					[da_axes addObject:@(i)];
+			}
+			mps_a = [graph reductionSumWithTensor:mps_a axes:da_axes name:nil];
+			[resultTensors addObject:mps_a];
+
+				if (h) {
+					MPSGraphTensor* mps_h = mps_g;
+					if (p != 1)
+					{
+						MPSGraphTensor* mps_p = [graph constantWithScalar:p dataType:ccv_nnc_mps_datatype(g->info.datatype)];
+						mps_h = [graph multiplicationWithPrimaryTensor:mps_h secondaryTensor:mps_p name:nil];
+					}
+					mps_h = [graph multiplicationWithPrimaryTensor:mps_h secondaryTensor:mps_b2 name:nil];
+
+					NSMutableArray<NSNumber*>* dh_axes = [NSMutableArray new];	
+					const int h_nd = ccv_nnc_tensor_nd(h->info.dim);
+					for (int i = 0; i < h_nd; i++) {
+						if (h->info.dim[i] != g->info.dim[i])
+							[dh_axes addObject:@(i)];
+					}
+					mps_h = [graph reductionSumWithTensor:mps_h axes:dh_axes name:nil];
+					[resultTensors addObject:mps_h];
+				}
+				
+			} else {
+				// empty mps_g for target broadcast shape with 1.0 for each elements	 
+				MPSGraphTensor* mps_g = [graph constantWithScalar:1.0 shape:mps_g_shape dataType:ccv_nnc_mps_datatype(a->info.datatype)];
```

Comment:
my workaround for broadcast... though it passed tests.. may need double check if this is acceptable..

Reply:
What's the workaround? I thought it is fine you just create a constant directly (with shape [1])? What's the error.

Reply:
right constant directly (with shape [1]), just not sure if this is common for broadcast shape implementation...

Reply:
Yeah, you just need to understand numpy broadcast semantics (which is what's implemented in MPSGraph).

a + b always first padding left until they align (assuming a is [6], b is [3, 4, 5, 6]): then first it is padded to [1, 1, 1, 6], then for every 1 dimension, it will broadcast to the value of b, if they don't match (a.dim[i] != b.dim[i] && a.dim[i] != 1 && b.dim[i] != 1), it will fail.